design.md

```diff
@@ -0,0 +1,221 @@
+
+1. gRPC API to start/stop/get status/stream output of a running process.
+1. Use mTLS for authentication.
+1. Has a simple authentication scheme.
+
+## Client
+
+1. CLI connects to worker service to start, stop, get status, and stream output of a job.
+2. Multiple clients can connect to the service.
+
+
+# Design
+
+## Worker Library and Server
+
+The server will use a library to: 
+
+1. Start/stop and query jobs. Starting a job returns a UUID which is the job id. This ID can be used to do further 
+   actions on the job. 
+2. Provide streaming of job logs.
+
+Service should be able to handle multiple clients. 
+
+### cgroups
+
+`cgroups v2` will be used for resource control. 
+
+- Each job will run in its own cgroup.
+- Resource usage will be isolated per job.
+- There will be job_id grouping. for example: 
+  - /sys/fs/cgroup/teleport/<job_id> -- the job_id is the UUID provided by the server when the job was created.
+
+The hardcoded resource limits are below. For this challenge, I will assume the jobs are relatively light: 
+- CPU: 0.5 
+- Memory: 256MB
+- IO: 1MB/s
+
+#### Placing jobs in cgroups
+
+- os.Open opens /sys/fs/cgroup/teleport/<job_id>, resulting in a file descriptor.
+- When running exec.Cmd, use UseCgroupFD and CgroupFD to specify that the process should be placed in a cgroup.
+
+
+#### Tradesoffs and the Production environment
+- It is preferable to dynamically configure resource limits based on job and system state. A first step in doing 
+  this is providing a way to adjust resource limits when jobs are created.
+- In production, we probably want orchestrators, like systemd, which manages cgroups. 
+
+## API
+
+### Using gRPC for communication. 
+
+- Start: Starts a job with hardcoded resource limits.
+- Stop: Stops a job.
+- GetStatus: Fetches status of a job
+- GetLogs: Stream logs of a running job.
+
+```
+service TeleportService {
+    rpc Start (StartRequest) returns (StartResponse);
+    rpc Stop (StopRequest) returns (StopResponse);
+    rpc GetStatus (GetStatusRequest) returns (GetStatusResponse);
+    rpc GetLogs (GetLogsRequest) returns (stream GetLogsResponse);
+}
+
+message StartRequest {
+    string command = 1;
+    repeated string args = 2;
+}
+
+message StartResponse {
+    string job_id = 1; // a job UUID
+}
+
+message StopRequest {
+    string job_id = 1;
+}
+
+message StopResponse {
+    // ideally, add a boolean to see if stopping was successful
+    // maybe even a message. But not necessary for the challenge.
+}
+
+message GetStatusRequest {
+    string job_id = 1;
+}
+
+message GetStatusResponse {
+    Status status = 1;
+}
+
+enum Status{
+    STATUS_RUNNING = 1;
+    STATUS_STOPPED = 2; // User stopped the job
+    STATUS_EXITED = 3;  // Job exited
+    STATUS_UNKNOWN = 4; // For anything we don't recognize
+}
+
+message LogsRequest {
+  string job_id = 1;
+}
+
+message LogsResponse {
+  bytes data = 1; // log data
+}
+```
+
+### SSL
+
+And API will use TLs 1.3, which is the latest version. 
```

Comment:
Can you provide justification for choosing TLS 1.3, and which ciphers will be used